File: python/src/subscriptions.py
import asyncio
import json

import websockets

from candlesticks import (
    BitfinexCandlestick,
    BitfinexCandlestickSnapshotData,
    BinanceCandlestickData,
    BitfinexCandlestickUpdateData,
)


async def subscribe_binance_candles(queue: asyncio.Queue):
    uri = "wss://stream.binance.com:9443/ws/btcusdt@kline_1m"
    async with websockets.connect(uri) as websocket:
        while True:
            msg: str = await websocket.recv()
            candle_obj = BinanceCandlestickData.model_validate_json(msg)
            await queue.put(candle_obj.k)


async def subscribe_bitfinex_candles(queue: asyncio.Queue):
    uri = "wss://api-pub.bitfinex.com/ws/2"
    async with websockets.connect(uri) as websocket:
        connection_params = (
            '{"event":"subscribe","channel":"candles","key":"trade:1m:tBTCUSD"}'
        )
        await websocket.send(connection_params)
        _ = await websocket.recv()
        _ = await websocket.recv()

        msg: str = await websocket.recv()

        obj = json.loads(msg)
        candles = BitfinexCandlestickSnapshotData(
            channel_id=obj[0],
            data=[
                BitfinexCandlestick(
                    mts=x[0],
                    open=x[1],
                    close=x[2],
                    high=x[3],
                    low=x[4],
                    volume=x[5],
                )
                for x in obj[1]
            ],
        )
        for candle in candles.data:
            await queue.put(candle)

        while True:
            msg: str = await websocket.recv()
            if "hb" in msg:
                continue
            obj = json.loads(msg)
            candle_obj = BitfinexCandlestickUpdateData(
                channel_id=obj[0],
                data=BitfinexCandlestick(
                    mts=obj[1][0],
                    open=obj[1][1],
                    close=obj[1][2],
                    high=obj[1][3],
                    low=obj[1][4],
                    volume=obj[1][5],
                ),
            )
            await queue.put(candle_obj)

File: rust/src/subscriptions.rs
use futures_util::SinkExt;
use futures_util::stream::StreamExt;
use tokio::sync::{broadcast, mpsc};
use tokio::time::Duration;
use tokio_tungstenite::connect_async;
use tokio_tungstenite::tungstenite::protocol::Message;

use crate::candlesticks::*;

pub async fn subscribe_binance_candles(
    sender: mpsc::Sender<BinanceCandlestick>,
    mut exit_receiver: broadcast::Receiver<()>, // Changed the type of exit_receiver
) {
    let url = "wss://stream.binance.com:9443/ws/btcusdt@kline_5m";
    let (mut ws_stream, _) = connect_async(url)
        .await
        .expect("Failed to connect to Binance");
    let mut interval = tokio::time::interval(Duration::from_secs(1));

    loop {
        tokio::select! {
            _ = exit_receiver.recv() => {
                println!("Closing Binance WebSocket");
                let _ = ws_stream.close(None).await;
                break;
            }
            _ = interval.tick() => {
                if let Some(msg) = ws_stream.next().await {
                    if let Ok(msg) = msg {
                        if let Message::Text(data) = msg {
                            let candlestick = serde_json::from_str::<BinanceCandlestickData>(&data).expect("Couldn't parse data");
                            sender.send(candlestick.k).await.expect("Failed to send to channel");
                        }
                    }
                }
            }
        }
    }
}

pub async fn subscribe_bitfinex_candles(
    sender: mpsc::Sender<BitfinexCandlestick>,
    mut exit_receiver: broadcast::Receiver<()>,
) {
    let url = "wss://api-pub.bitfinex.com/ws/2";
    let (mut ws_stream, _) = connect_async(url)
        .await
        .expect("Failed to connect to Bitfinex");
    let _info_msg = ws_stream.next().await;
    let _ = ws_stream.send(Message::from("{\"event\":\"subscribe\",\"channel\":\"candles\",\"key\":\"trade:1m:tBTCUSD\"}")).await;
    let _subscription_msg = ws_stream.next().await;
    if let Some(Ok(snapshot_msg)) = ws_stream.next().await {
        if let Message::Text(data) = snapshot_msg {
            let snapshot = serde_json::from_str::<BitfinexCandlestickSnapshotData>(&data).expect("Couldn't parse data");
            for candlestick in snapshot.candle_data.iter() {
                sender.send(candlestick.clone()).await.expect("Failed to send to channel");
            }
        }
    };


    let mut interval = tokio::time::interval(Duration::from_secs(1));

    loop {
        tokio::select! {
            _ = exit_receiver.recv() => {
                println!("Closing Bitfinex WebSocket");
                let _ = ws_stream.close(None).await;
                break;
            }
            _ = interval.tick() => {
                if let Some(msg) = ws_stream.next().await {
                    if let Ok(msg) = msg {
                        if let Message::Text(data) = msg {
                            if data.contains("\"event\"") || data.contains("hb") { continue }
                            let candlestick = serde_json::from_str::<BitfinexCandlestickUpdateData>(&data).expect("Couldn't parse data");
                            sender.send(candlestick.candle_data.clone()).await.expect("Failed to send to channel");
                        }
                    }
                }
            }
        }
    }
}